### PR TITLE
fix: save crash reports locally when uploadToServer: false on linux

### PR DESCRIPTION
**Pull request by deepak1556**
#### Description of Change

Fixes https://github.com/electron/electron/issues/24714

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [ ] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).
- [x] This is **NOT A BREAKING CHANGE**. Breaking changes may not be merged to master until 11-x-y is branched.

#### Release Notes

Notes: save crash reports locally when uploadToServer: false on linux

### CODE REVIEW

File: shell/app/electron_crash_reporter_client.cc [hunk 1]
```diff
@@ -147,7 +148,14 @@ bool ElectronCrashReporterClient::GetCrashDumpLocation(
 #else
 bool ElectronCrashReporterClient::GetCrashDumpLocation(
     base::FilePath* crash_dir) {
-  return base::PathService::Get(electron::DIR_CRASH_DUMPS, crash_dir);
+  bool result = base::PathService::Get(electron::DIR_CRASH_DUMPS, crash_dir);
+  {
+    base::ThreadRestrictions::ScopedAllowIO allow_io;
+    if (result && !base::PathExists(*crash_dir)) {
+      return base::CreateDirectory(*crash_dir);
```
Comment (deepak1556) [inline shell/app/electron_crash_reporter_client.cc:155]: Breakpad path doesn't seem to initialize the crashes directory and without it crash dumping fails.

File: spec-main/api-crash-reporter-spec.ts [hunk 2]
```diff
@@ -485,24 +485,38 @@ ifdescribe(!isLinuxOnArm && !process.mas && !process.env.DISABLE_CRASH_REPORTER_
       }
     }
 
-    for (const crashingProcess of ['main', 'renderer', 'sandboxed-renderer', 'node']) {
-      // TODO(nornagon): breakpad on linux disables itself when uploadToServer
-      // is false, so we should figure out a different way to test the crash
-      // dump dir on linux.
-      ifdescribe(process.platform !== 'linux')(`when ${crashingProcess} crashes`, () => {
+    const processList = process.platform === 'linux' ? ['main', 'renderer', 'sandboxed-renderer']
+      : ['main', 'renderer', 'sandboxed-renderer', 'node'];
```
Comment (deepak1556) [inline spec-main/api-crash-reporter-spec.ts:489]: fix for node process on linux is non-trivial, filed https://github.com/electron/electron/issues/24779 for follow-up.

File: spec-main/api-crash-reporter-spec.ts [hunk 3]
```diff
@@ -485,24 +485,38 @@ ifdescribe(!isLinuxOnArm && !process.mas && !process.env.DISABLE_CRASH_REPORTER_
...
           const newFileAppeared = waitForNewFileInDir(reportsDir);
           crash(crashingProcess, remotely);
           const newFiles = await newFileAppeared;
           expect(newFiles.length).to.be.greaterThan(0);
-          expect(newFiles[0]).to.match(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\.dmp$/);
+          if (process.platform === 'linux') {
+            if (crashingProcess === 'main') {
+              expect(newFiles[0]).to.match(/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{8}-[0-9a-f]{8}\.dmp$/);
+            } else {
+              const process = crashingProcess !== 'sandboxed-renderer' ? crashingProcess : 'renderer';
```
Comment (nornagon) [inline spec-main/api-crash-reporter-spec.ts:513]:
```suggestion
              const process = crashingProcess === 'sandboxed-renderer' ? 'renderer' : crashingProcess;
```

a little easier to read

File: shell/app/electron_crash_reporter_client.cc [hunk 4]
```diff
@@ -159,7 +167,7 @@ bool ElectronCrashReporterClient::GetCrashMetricsLocation(
 #endif  // OS_MACOSX || OS_LINUX
 
 bool ElectronCrashReporterClient::IsRunningUnattended() {
-  return false;
+  return !collect_stats_consent_;
```
[the remark this answers is not in the document]
Comment (deepak1556) [reply shell/app/electron_crash_reporter_client.cc:170]: yup the crash dumping to disk for the browser process relies on this https://source.chromium.org/chromium/chromium/src/+/master:components/crash/core/app/breakpad_linux.cc;l=2068

[the remark this answers is not in the document]
Comment (nornagon) [reply shell/app/electron_crash_reporter_client.cc:170]: aha, this comment explains why the redundancy: https://source.chromium.org/chromium/chromium/src/+/master:components/crash/core/app/crashpad.cc;l=251-255;drc=e4696af2b851af0b99601ec463ef11209036c2bd

**Comment by deepak1556** [comment]
Failing tests are unrelated, merging.

**Pull request merged**